stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1027,9 +1027,13 @@ julia> tr(A)
 5
 ```
 """
-function tr(A::AbstractMatrix)
-    checksquare(A)
-    sum(diag(A))
+function tr(A::AbstractMatrix{T}) where T
```

Comment:
Maybe needs
```suggestion
function tr(A::AbstractMatrix{T}) where T
    require_one_based_indexing(A)
```
to index via `1:n`? I think this isn’t a new restriction, since it comes up in [the generic `diag` method](https://github.com/JuliaLang/julia/blob/f613b551009a7a9dbe46235929099f2ecd28bed1/stdlib/LinearAlgebra/src/dense.jl#L221), but it does mean if a non-one-based matrix has defined a `diag` method but not a `tr` method then `tr` will stop working, which seems bad.

Reply:
Thanks for catching. We could just iterate over `diagind(A)`?

Reply:
That seems better to me, though there's still the issue that if a type defined a custom `diag` but not `diagind` or `tr` then this fallback `tr` method will stop working for it. I'm not really sure what's allowed or not with this kind of change, maybe someone else will know

Reply:
cc @simeonschaub @andreasnoack 
btw this was a tiny bit faster than `sum(view(A, 1:size(A, 1)+1:length(A)))` for sparse arrays, e.g.:
```julia
julia> using LinearAlgebra, SparseArrays
julia> n = 50_000; A = sprand(n, n, 0.01);
julia> @time tr(A);
  0.003509 seconds (1 allocation: 16 bytes)
julia> @time sum(view(A, 1:size(A, 1)+1:length(A)));
  0.004263 seconds (6 allocations: 224 bytes)
julia> @time sum(view(A, diagind(A)));
  0.007038 seconds (3 allocations: 176 bytes)
```

Reply:
@ericphanson wouldn't adding `diagind(x)=eachindex(diag(x))` fix this problem?

Reply:
@oscardssmith isn't this `1:length(diag(x))`?

Reply:
it almost certainly is, but using `eachindex` means that if someone defines `diag` to return an `OffsetArray` it will still work.

Reply:
> @ericphanson wouldn't adding `diagind(x)=eachindex(diag(x))` fix this problem?

This seems like it could be circular in some cases, e.g. if a 1-based `AbstractMatrix` does not define a specialized `diag` method, then currently what happens is it hits the `diag` fallback, which calls the generic `diagind` method, and everything works. But if we change the `diagind` fallback to something like `diagind(x)=eachindex(diag(x))`  then this is circular in that case.

It seems like we'd want that fallback *only* for non-1-based `AbstractMatrix`'s, but we don't have a trait or such for that AFAIK, so we can't dispatch it correctly at compile time. Maybe we could do a runtime check and in the 1-based case use the current method, and in the non-1-based case use `eachindex(diag(x))` as a new fallback?

Also, doesn't that re-introduce the allocations this PR is hoping to avoid? Or can Julia elide the `diag` allocation?

Reply:
I'm probably missing something obvious here- but I still don't see how this `diagind` gives us what we want. If what you mean is to iterate over `eachindex(diag(x))`, then as @ericphanson mentioned this re-introduces the allocations. If what you mean to define this `diagind` as a separate fallback, then this doesn't match the definition of `diagind`, which should return the indices of the diagonal of `x` inside `x`.